# A Complete guide to Data Analysis using Python

# Import Packages

In [57]:

# Import pandas package
import pandas as pd

# Pandas - Series

A Series can be created from a list of values, arrays and dictionaries.

# Create series from lists


Let's create a new series using a list of values

In [58]:
s1 = pd.Series([10,20,30,40,50])

In [59]:
print("The series values are:", s1.values)
print("The index values are:", s1.index.values)

The series values are: [10 20 30 40 50]
The index values are: [0 1 2 3 4]


The index values are generated by default, but we can also define custom indexes at the time of creating series.

Below is a Series of marks and the associated subjects can be defined as custom row indexes

In [60]:
s2 = pd.Series([80,93,78,85,97], index=['English','Science','Social','Tamil','Maths'])

In [61]:
print("The Marks obtained by student are", s2)

The Marks obtained by student are English    80
Science    93
Social     78
Tamil      85
Maths      97
dtype: int64


Both the series values and index can have name of its own and its defined as follows

In [45]:
s2.name = 'Student Marks'
s2.index.name = 'Subject'

In [46]:
print("The Marks obtained by student - subjectwise:\n", s2)

The Marks obtained by student - subjectwise:
 Subject
English    80
Science    93
Social     78
Tamil      85
Maths      97
Name: Student Marks, dtype: int64


# Subsetting the series

Subsetting the series
Values inside the series can be accessed using the [ ] square bracket slicing operation

*  Series can be sliced using the default index
*  Series can also be sliced using string indices

In [47]:
s2[0]

80

In [48]:
# slicing using default integer index
s1[1:4]

1    20
2    30
3    40
dtype: int64

In [62]:

# Slicing using string index
s2['Tamil']

85

# Creating Series from Dictionary

* A Dictionary is a data structure which maps keys to a set of values.
* A dictionary contains key-value pairs
* A Series is similar to a dictionary in a way that it maps given indexes to a set of values
Let's see how to create a Series from a dictionary

I have a dictionary that represents fruits and prices. Let's create a series from this dictionary

In [67]:
dict_fruits = { 'Orange':80,
          'Apples':210,
          'Bananas':50,
         'Grapes':90,
         'Watermelon':70}

In [68]:
dict_fruits

{'Orange': 80, 'Apples': 210, 'Bananas': 50, 'Grapes': 90, 'Watermelon': 70}

In [69]:
# Lets convert this dictionary into a series
fruits = pd.Series(dict_fruits)

In [70]:
print("Fruits and prices\n", fruits)

Fruits and prices
 Orange         80
Apples        210
Bananas        50
Grapes         90
Watermelon     70
dtype: int64


In [71]:
# Slice the series and retrieve price of Grapes
print("The price per kg of grapes is:", fruits['Grapes'])

The price per kg of grapes is: 90


# Pandas - DataFrame

# 1. Create DataFrame from Series object

Let's create a DataFrame from the Marks series we created in the above step.

In [72]:
df_marks = pd.DataFrame(s2, columns=['Student1'])

In [73]:
print("The dataframe created from series is\n",df_marks)

The dataframe created from series is
          Student1
English        80
Science        93
Social         78
Tamil          85
Maths          97


# 2. Create DataFrame from Dictionary

In [75]:
# Create Height series (in feet)
height = pd.Series([5.3, 6.2,5.8,5.0,5.5], index=['Person 1','Person 2','Person 3','Person 4','Person 5'])

# Create Weight Series (in kgs)
weight = pd.Series([65,89,75,60,59], index=['Person 1','Person 2','Person 3','Person 4','Person 5'])

In [77]:
#Create dataframe
df_person = pd.DataFrame({'height': height, 'weight': weight})
print("The Person table details are:\n", df_person)

The Person table details are:
           height  weight
Person 1     5.3      65
Person 2     6.2      89
Person 3     5.8      75
Person 4     5.0      60
Person 5     5.5      59


# 3. Create DataFrame by importing data from file

Pandas is extremely useful and comes handy when want to load data from various file formats like csv, excel, json etc.

For this analysis we will load data from a csv file.

# Data Analysis using Pandas

We will use 'IMDB-Movie-Data' opensource data to learn more about pandas. This dataset contains 7787 rows of data in 12 columns. It contains data about movies and shows on netflix.

Lets dive deep in to this data for some fun and interesting analysis. Put on your hats of a Data Analyst/ Data Scientist and let's GET.SET.GO

We will read data from csv files and peform some basic analysis on the dataset

We will perform these basic operations

1. Read data
2. View the data
3. Understand some basic information about the data
4. Data Selection - Indexing and slicing operations
5. Data Selection - Conditional filtering
6. Groupby operation
7. Sorting operation
8. Dealing with missing values
9. Dropping columns
10. Apply( ) function
11. Pivot tables

# 1. Read Data

In [80]:
# Read data from .csv file
data = pd.read_csv("D:\ML-Data-Sets\IMDB-Movie-Data.csv")
# Read data with specified explicit index.
# We will use this later in our analysis
data_indexed = pd.read_csv("D:\ML-Data-Sets\IMDB-Movie-Data.csv", index_col="Title")

In [81]:
#Lets first understand the basic information about this data

data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rank                1000 non-null   int64  
 1   Title               1000 non-null   object 
 2   Genre               1000 non-null   object 
 3   Description         1000 non-null   object 
 4   Director            1000 non-null   object 
 5   Actors              1000 non-null   object 
 6   Year                1000 non-null   int64  
 7   Runtime (Minutes)   1000 non-null   int64  
 8   Rating              1000 non-null   float64
 9   Votes               1000 non-null   int64  
 10  Revenue (Millions)  872 non-null    float64
 11  Metascore           936 non-null    float64
dtypes: float64(3), int64(4), object(5)
memory usage: 93.9+ KB


As you can see, the info( ) method gives all the details about this dataframe like columns, number of observations and the datatype of these columns.

info( ) method comes in handy when we first explore the data

Let's do a quick preview of the data by using head( ) and tail( ) methods

* head( ) ---> returns the top 5 rows in the dataset by default head( ) can also take the number of rows to be viewed as a parameter

* tail( ) ---> returns the bottom 5 rows in the dataset by default tail( ) can also take the number of rows as an optional parameter

In [82]:
# Preview top 5 rows using head()
data.head()

Rank                    Title                     Genre  \
0     1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1     2               Prometheus  Adventure,Mystery,Sci-Fi   
2     3                    Split           Horror,Thriller   
3     4                     Sing   Animation,Comedy,Family   
4     5            Suicide Squad  Action,Adventure,Fantasy   

                                         Description              Director  \
0  A group of intergalactic criminals are forced ...            James Gunn   
1  Following clues to the origin of mankind, a te...          Ridley Scott   
2  Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3  In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4  A secret government agency recruits some of th...            David Ayer   

                                              Actors  Year  Runtime (Minutes)  \
0  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014                121   
1  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012                124   
2  James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016                117   
3  Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016                108   
4  Will Smith, Jared Leto, Margot Robbie, Viola D...  2016                123   

   Rating   Votes  Revenue (Millions)  Metascore  
0     8.1  757074              333.13       76.0  
1     7.0  485820              126.46       65.0  
2     7.3  157606              138.12       62.0  
3     7.2   60545              270.32       59.0  
4     6.2  393727              325.02       40.0

data.tail()

The column names in the data can be viewed by using 'columns'

In [84]:
data.columns

Index(['Rank', 'Title', 'Genre', 'Description', 'Director', 'Actors', 'Year',
       'Runtime (Minutes)', 'Rating', 'Votes', 'Revenue (Millions)',
       'Metascore'],
      dtype='object')

The shape of the dataset can be viewed by using 'shape'

In [85]:
data.shape

(1000, 12)

This function tells us that there are 1000 rows and 12 columns in the dataset

Let's use describe( ) method to understand the numerical attributes in the data

<b>describe( )</b> shows the basic statistical summaries of numerical attributes in the data.

In [86]:
data.describe()

Rank         Year  Runtime (Minutes)       Rating         Votes  \
count  1000.000000  1000.000000        1000.000000  1000.000000  1.000000e+03   
mean    500.500000  2012.783000         113.172000     6.723200  1.698083e+05   
std     288.819436     3.205962          18.810908     0.945429  1.887626e+05   
min       1.000000  2006.000000          66.000000     1.900000  6.100000e+01   
25%     250.750000  2010.000000         100.000000     6.200000  3.630900e+04   
50%     500.500000  2014.000000         111.000000     6.800000  1.107990e+05   
75%     750.250000  2016.000000         123.000000     7.400000  2.399098e+05   
max    1000.000000  2016.000000         191.000000     9.000000  1.791916e+06   

       Revenue (Millions)   Metascore  
count          872.000000  936.000000  
mean            82.956376   58.985043  
std            103.253540   17.194757  
min              0.000000   11.000000  
25%             13.270000   47.000000  
50%             47.985000   59.500000  
75%            113.715000   72.000000  
max            936.630000  100.000000

<b>Some Insights from description of data</b>

* The min and max values in 'Year' depict the minimum and maximum release years. We can see that the dataset contains movies from 2006 to 2016.

* The average rating for the movies in this dataset is about 6.7 and the mininum rating is 1.9 and the maximum rating is 9.0

* The maximum revenue earned by a movie is 936.6 millions

In [89]:
# check null values
data.isnull().sum()

Rank                    0
Title                   0
Genre                   0
Description             0
Director                0
Actors                  0
Year                    0
Runtime (Minutes)       0
Rating                  0
Votes                   0
Revenue (Millions)    128
Metascore              64
dtype: int64

<b>isnull()</b> function shows how many missing values are in the dataset with respect to different columns.

Here 'Revenue' and 'Metascore' columns has null values

# Data Selection - Indexing and Slicing


Extracting data from Pandas column is similar to Series. The column label is used to extract data from the column

<b>Extracting data by columns</b>

In [90]:
# Extract data as series
genre = data['Genre']

In [91]:
genre 

0       Action,Adventure,Sci-Fi
1      Adventure,Mystery,Sci-Fi
2               Horror,Thriller
3       Animation,Comedy,Family
4      Action,Adventure,Fantasy
                 ...           
995         Crime,Drama,Mystery
996                      Horror
997         Drama,Music,Romance
998            Adventure,Comedy
999       Comedy,Family,Fantasy
Name: Genre, Length: 1000, dtype: object

This will return the 'genre' column as a series.

If we want to retrive it as a dataframe, then indexing must be done using double square brackets

In [92]:
type(genre)

pandas.core.series.Series

In [93]:
# Extract data as dataframe
data[['Genre']]

Genre
0     Action,Adventure,Sci-Fi
1    Adventure,Mystery,Sci-Fi
2             Horror,Thriller
3     Animation,Comedy,Family
4    Action,Adventure,Fantasy
..                        ...
995       Crime,Drama,Mystery
996                    Horror
997       Drama,Music,Romance
998          Adventure,Comedy
999     Comedy,Family,Fantasy

[1000 rows x 1 columns]

In [94]:
type(data[["Genre"]])

pandas.core.frame.DataFrame

If we want to extract multiple columns from the data, simply add the column names to the list

In [95]:
some_cols = data[['Title','Genre','Actors','Director','Rating']]

In [96]:
some_cols.head()

Title                     Genre  \
0  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1               Prometheus  Adventure,Mystery,Sci-Fi   
2                    Split           Horror,Thriller   
3                     Sing   Animation,Comedy,Family   
4            Suicide Squad  Action,Adventure,Fantasy   

                                              Actors              Director  \
0  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...            James Gunn   
1  Noomi Rapace, Logan Marshall-Green, Michael Fa...          Ridley Scott   
2  James McAvoy, Anya Taylor-Joy, Haley Lu Richar...    M. Night Shyamalan   
3  Matthew McConaughey,Reese Witherspoon, Seth Ma...  Christophe Lourdelet   
4  Will Smith, Jared Leto, Margot Robbie, Viola D...            David Ayer   

   Rating  
0     8.1  
1     7.0  
2     7.3  
3     7.2  
4     6.2

<b>Extract data by row index</b>

We already have a dataset with 'title' as index and we can subset that dataframe using <b>loc</b>

If we want to slice data from specific rows, loc and iloc can be used

<b>loc</b>

Here indexing is done based on explicit index - locates by name.
It can take string indexes to retrive data from the specified rows
<b>iloc</b>

* Here indexing is done based on Python's numerical index - locates by integer index.
* This works only with integer indexes to retrieve data from specified rows

In [106]:
data_indexed.head()

Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   

                                     Director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   

                                                                    Actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
Guardians of the Galaxy  2014                121     8.1  757074   
Prometheus               2012                124     7.0  485820   
Split                    2016                117     7.3  157606   
Sing                     2016                108     7.2   60545   
Suicide Squad            2016                123     6.2  393727   

                         Revenue (Millions)  Metascore  
Title                                                   
Guardians of the Galaxy              333.13       76.0  
Prometheus                           126.46       65.0  
Split                                138.12       62.0  
Sing                                 270.32       59.0  
Suicide Squad                        325.02       40.0

In [107]:
# Retrieving movie with title 'Sing'
data_indexed.loc['Suicide Squad']

Rank                                                                  5
Genre                                          Action,Adventure,Fantasy
Description           A secret government agency recruits some of th...
Director                                                     David Ayer
Actors                Will Smith, Jared Leto, Margot Robbie, Viola D...
Year                                                               2016
Runtime (Minutes)                                                   123
Rating                                                              6.2
Votes                                                            393727
Revenue (Millions)                                               325.02
Metascore                                                          40.0
Name: Suicide Squad, dtype: object

In [108]:
data_indexed.loc[['Suicide Squad']][['Genre','Actors','Director','Rating','Revenue (Millions)']]

Genre  \
Title                                     
Suicide Squad  Action,Adventure,Fantasy   

                                                          Actors    Director  \
Title                                                                          
Suicide Squad  Will Smith, Jared Leto, Margot Robbie, Viola D...  David Ayer   

               Rating  Revenue (Millions)  
Title                                      
Suicide Squad     6.2              325.02

<b>
Retrieving data with row indices</b>

In [109]:
# Pick rows from indexes 10-49
data.iloc[10:15]

Rank                                    Title                       Genre  \
10    11  Fantastic Beasts and Where to Find Them    Adventure,Family,Fantasy   
11    12                           Hidden Figures     Biography,Drama,History   
12    13                                Rogue One     Action,Adventure,Sci-Fi   
13    14                                    Moana  Animation,Adventure,Comedy   
14    15                                 Colossal         Action,Comedy,Drama   

                                          Description         Director  \
10  The adventures of writer Newt Scamander in New...      David Yates   
11  The story of a team of female African-American...   Theodore Melfi   
12  The Rebel Alliance makes a risky move to steal...   Gareth Edwards   
13  In Ancient Polynesia, when a terrible curse in...     Ron Clements   
14  Gloria is an out-of-work party girl forced to ...  Nacho Vigalondo   

                                               Actors  Year  \
10  Eddie Redmayne, Katherine Waterston, Alison Su...  2016   
11  Taraji P. Henson, Octavia Spencer, Janelle Mon...  2016   
12  Felicity Jones, Diego Luna, Alan Tudyk, Donnie...  2016   
13  Auli'i Cravalho, Dwayne Johnson, Rachel House,...  2016   
14  Anne Hathaway, Jason Sudeikis, Austin Stowell,...  2016   

    Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  
10                133     7.5  232072              234.02       66.0  
11                127     7.8   93103              169.27       74.0  
12                133     7.9  323118              532.17       65.0  
13                107     7.7  118151              248.75       81.0  
14                109     6.4    8612                2.87       70.0

In [110]:
data.iloc[10:15][['Title','Rating','Revenue (Millions)']]

Title  Rating  Revenue (Millions)
10  Fantastic Beasts and Where to Find Them     7.5              234.02
11                           Hidden Figures     7.8              169.27
12                                Rogue One     7.9              532.17
13                                    Moana     7.7              248.75
14                                 Colossal     6.4                2.87

# Data Selection - based on Conditional Filtering

In [111]:
import numpy as np


Let's try to pick the 'Title','Actors','Director' for movies where the rating is minimum

In [119]:
np.where(data['Rating'] == 1.9)
    

(array([829], dtype=int64),)

In [114]:
rating_min = np.where(data['Rating'] == min(data['Rating']))
rating_min

(array([829], dtype=int64),)

In [120]:
data.iloc[rating_min][['Title','Actors','Director','Year','Rating']]

Title                                             Actors  \
829  Disaster Movie  Carmen Electra, Vanessa Lachey,Nicole Parker, ...   

            Director  Year  Rating  
829  Jason Friedberg  2008     1.9

What if we want to pick only movies that are released from 2010 to 2016, have a rating of less than 6.0 but topped in terms of revenue

In [154]:
data[((data['Year'] >= 2010) & (data['Year'] <= 2016)) & (data['Rating'] < 6.0) & (data['Revenue (Millions)'] > data['Revenue (Millions)'].quantile(0.95))] 

Rank                       Title                    Genre  \
941   942  The Twilight Saga: Eclipse  Adventure,Drama,Fantasy   

                                           Description     Director  \
941  As a string of mysterious killings grips Seatt...  David Slade   

                                                Actors  Year  \
941  Kristen Stewart, Robert Pattinson, Taylor Laut...  2010   

     Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  
941                124     4.9  192740              300.52       58.0

In [174]:
# Get unique values in rating column
data['Director'].unique()

array(['James Gunn', 'Ridley Scott', 'M. Night Shyamalan',
       'Christophe Lourdelet', 'David Ayer', 'Yimou Zhang',
       'Damien Chazelle', 'Sean Foley', 'James Gray', 'Morten Tyldum',
       'David Yates', 'Theodore Melfi', 'Gareth Edwards', 'Ron Clements',
       'Nacho Vigalondo', 'Chris Renaud', 'Mel Gibson', 'Paul Greengrass',
       'Garth Davis', 'Denis Villeneuve', 'Stephen Gaghan',
       'Kenneth Lonergan', 'Ben Young', 'Walt Dohrn', 'Roland Emmerich',
       'Dominique Abel', 'S.S. Rajamouli', 'Phillip Guzman', 'Jon Lucas',
       'Justin Kurzel', 'John Hamburg', 'Tom Ford', 'Bryan Singer',
       'Tim Miller', 'Paul W.S. Anderson', 'Anthony Russo',
       'Christopher Nolan', 'Scott Derrickson', 'Antoine Fuqua',
       'Patrick Read Johnson', 'Greg Tiernan', 'Barry Jenkins',
       'Shawn Burkett', 'John Lee Hancock', 'Ricardo de Montreuil',
       'Rob Marshall', 'John Madden', 'Scott Hicks', 'Justin Lin',
       'Sean Penn', 'J.J. Abrams', 'Anna Foerster', 'Garry Mar

In [164]:
data['Director'].value_counts()

Director
Ridley Scott          8
David Yates           6
M. Night Shyamalan    6
Paul W.S. Anderson    6
Michael Bay           6
                     ..
Lee Toland Krieger    1
Gillies MacKinnon     1
Peter Atencio         1
James Mangold         1
Scot Armstrong        1
Name: count, Length: 644, dtype: int64

# Groupby operations

* groupby is another interesting operation that can be peformed using Pandas groupby( ) method.
* This comes in handly when we want to apply aggregation operations and functions on a set of grouped data
Let's group the data by Director and see what's the revenue earned by each director

In [176]:
data.groupby('Director').sum()

Rank                                              Title  \
Director                                                                       
Aamir Khan            992                                   Taare Zameen Par   
Abdellatif Kechiche   312                                     La vie d'Adèle   
Adam Leon             784                                             Tramps   
Adam McKay           1910  The Big ShortStep BrothersThe Other GuysTallad...   
Adam Shankman        1460                              HairsprayRock of Ages   
...                   ...                                                ...   
Xavier Dolan         1588                MommyIt's Only the End of the World   
Yimou Zhang             6                                     The Great Wall   
Yorgos Lanthimos      479                              The LobsterKynodontas   
Zack Snyder           904  Batman v Superman: Dawn of Justice300WatchmenS...   
Zackary Adler         605                              The Rise of the Krays   

                                                                 Genre  \
Director                                                                 
Aamir Khan                                          Drama,Family,Music   
Abdellatif Kechiche                                      Drama,Romance   
Adam Leon                                               Comedy,Romance   
Adam McKay           Biography,Comedy,DramaComedyAction,Comedy,Crim...   
Adam Shankman                  Comedy,Drama,FamilyComedy,Drama,Musical   
...                                                                ...   
Xavier Dolan                                                DramaDrama   
Yimou Zhang                                   Action,Adventure,Fantasy   
Yorgos Lanthimos                    Comedy,Drama,RomanceDrama,Thriller   
Zack Snyder          Action,Adventure,Sci-FiAction,Fantasy,WarActio...   
Zackary Adler                                              Crime,Drama   

                                                           Description  \
Director                                                                 
Aamir Khan           An eight-year-old boy is thought to be a lazy ...   
Abdellatif Kechiche  Adèle's life is changed when she meets Emma, a...   
Adam Leon            A young man and woman find love in an unlikely...   
Adam McKay           Four denizens in the world of high-finance pre...   
Adam Shankman        Pleasantly plump teenager Tracy Turnblad teach...   
...                                                                ...   
Xavier Dolan         A widowed single mother, raising her violent s...   
Yimou Zhang          European mercenaries searching for black powde...   
Yorgos Lanthimos     In a dystopian near future, single people, acc...   
Zack Snyder          Fearing that the actions of Superman are left ...   
Zackary Adler        Two brothers unleash a psychotic reign of terr...   

                                                                Actors   Year  \
Director                                                                        
Aamir Khan           Darsheel Safary, Aamir Khan, Tanay Chheda, Sac...   2007   
Abdellatif Kechiche  Léa Seydoux, Adèle Exarchopoulos, Salim Kechio...   2013   
Adam Leon            Callum Turner, Grace Van Patten, Michal Vondel...   2016   
Adam McKay           Christian Bale, Steve Carell, Ryan Gosling, Br...   8039   
Adam Shankman        John Travolta, Queen Latifah, Nikki Blonsky,Mi...   4019   
...                                                                ...    ...   
Xavier Dolan         Anne Dorval, Antoine-Olivier Pilon, Suzanne Cl...   4030   
Yimou Zhang              Matt Damon, Tian Jing, Willem Dafoe, Andy Lau   2016   
Yorgos Lanthimos     Colin Farrell, Rachel Weisz, Jessica Barden,Ol...   4024   
Zack Snyder          Ben Affleck, Henry Cavill, Amy Adams, Jesse Ei...  10055   
Zackary Adler        Matt Vael, Simon Cotton, Kevin Leslie, Olivia ...   2015   




Here all of the data is grouped by 'Director' column and then aggregation 'sum' is applied.

Aggregation is applied to all of the numerical attributes in the data.

If we want to apply specific aggregation to only selected attribute, we can do so selecting the columns as described below.

In [177]:
data.groupby('Director')[['Rating']].mean()

Rating
Director                   
Aamir Khan             8.50
Abdellatif Kechiche    7.80
Adam Leon              6.50
Adam McKay             7.00
Adam Shankman          6.30
...                     ...
Xavier Dolan           7.55
Yimou Zhang            6.10
Yorgos Lanthimos       7.20
Zack Snyder            7.04
Zackary Adler          5.10

[644 rows x 1 columns]

# Sorting

* Sorting is one of the most important functions of pandas which is heavily used for data analysis
In the above example, where we have listed the average rating for each Director, if we want to sort them from highly rated to lowest, we can perform the sorting operation

In [178]:
data.groupby('Director')[['Rating']].mean().sort_values(['Rating'], ascending=False).head()

Rating
Director                 
Nitesh Tiwari        8.80
Christopher Nolan    8.68
Olivier Nakache      8.60
Makoto Shinkai       8.60
Aamir Khan           8.50

We can see that Director 'Nitesh Tiwari' has the highest average rating in this dataset

Now if we want to see which movies had the highest revenue and the highest rating, we can do so by following this

In [181]:
data.sort_values(['Revenue (Millions)','Rating'], ascending=False).head(5)

Rank                                       Title  \
50    51  Star Wars: Episode VII - The Force Awakens   
87    88                                      Avatar   
85    86                              Jurassic World   
76    77                                The Avengers   
54    55                             The Dark Knight   

                       Genre  \
50  Action,Adventure,Fantasy   
87  Action,Adventure,Fantasy   
85   Action,Adventure,Sci-Fi   
76             Action,Sci-Fi   
54        Action,Crime,Drama   

                                          Description           Director  \
50  Three decades after the defeat of the Galactic...        J.J. Abrams   
87  A paraplegic marine dispatched to the moon Pan...      James Cameron   
85  A new theme park, built on the original site o...    Colin Trevorrow   
76  Earth's mightiest heroes must come together an...        Joss Whedon   
54  When the menace known as the Joker wreaks havo...  Christopher Nolan   

                                               Actors  Year  \
50  Daisy Ridley, John Boyega, Oscar Isaac, Domhna...  2015   
87  Sam Worthington, Zoe Saldana, Sigourney Weaver...  2009   
85  Chris Pratt, Bryce Dallas Howard, Ty Simpkins,...  2015   
76  Robert Downey Jr., Chris Evans, Scarlett Johan...  2012   
54  Christian Bale, Heath Ledger, Aaron Eckhart,Mi...  2008   

    Runtime (Minutes)  Rating    Votes  Revenue (Millions)  Metascore  
50                136     8.1   661608              936.63       81.0  
87                162     7.8   935408              760.51       83.0  
85                124     7.0   455169              652.18       59.0  
76                143     8.1  1045588              623.28       69.0  
54                152     9.0  1791916              533.32       82.0

Here sorting operation is performed on Revenue followed by Rating

# Dealing with missing values

Pandas has isnull( ) and notnull( ) for detecting null values in a dataframe

Lets see how to use these

In [184]:

# To see if there are null values in the whole data
data.isnull()

Rank  Title  Genre  Description  Director  Actors   Year  \
0    False  False  False        False     False   False  False   
1    False  False  False        False     False   False  False   
2    False  False  False        False     False   False  False   
3    False  False  False        False     False   False  False   
4    False  False  False        False     False   False  False   
..     ...    ...    ...          ...       ...     ...    ...   
995  False  False  False        False     False   False  False   
996  False  False  False        False     False   False  False   
997  False  False  False        False     False   False  False   
998  False  False  False        False     False   False  False   
999  False  False  False        False     False   False  False   

     Runtime (Minutes)  Rating  Votes  Revenue (Millions)  Metascore  
0                False   False  False               False      False  
1                False   False  False               False      False  
2                False   False  False               False      False  
3                False   False  False               False      False  
4                False   False  False               False      False  
..                 ...     ...    ...                 ...        ...  
995              False   False  False                True      False  
996              False   False  False               False      False  
997              False   False  False               False      False  
998              False   False  False                True      False  
999              False   False  False               False      False  

[1000 rows x 12 columns]

In [185]:
# To check null values row-wise
data.isnull().sum()

Rank                    0
Title                   0
Genre                   0
Description             0
Director                0
Actors                  0
Year                    0
Runtime (Minutes)       0
Rating                  0
Votes                   0
Revenue (Millions)    128
Metascore              64
dtype: int64

Here we know that 'Revenue (Millions)' and 'Metascore' are two columns where there are null values

As we have seen null values in data, we can either choose to drop those or impute these values

# Dropping null values

1. Dropping can be done either by rows or by columns depending on our need
2. drop( ), dropna( ) are some functions used to drop null values
3. When we drop values, either the complete column or row is dropped and not the single values
Let's say I want to drop the rows where Metascore is null

Let's say I want to drop the rows where Metascore is null

In [203]:
# Drops all rows containing missing data
data.dropna(axis=0)

Rank                     Title                     Genre  \
0       1   Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1       2                Prometheus  Adventure,Mystery,Sci-Fi   
2       3                     Split           Horror,Thriller   
3       4                      Sing   Animation,Comedy,Family   
4       5             Suicide Squad  Action,Adventure,Fantasy   
..    ...                       ...                       ...   
993   994  Resident Evil: Afterlife   Action,Adventure,Horror   
994   995                 Project X                    Comedy   
996   997           Hostel: Part II                    Horror   
997   998    Step Up 2: The Streets       Drama,Music,Romance   
999  1000                Nine Lives     Comedy,Family,Fantasy   

                                           Description              Director  \
0    A group of intergalactic criminals are forced ...            James Gunn   
1    Following clues to the origin of mankind, a te...          Ridley Scott   
2    Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3    In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4    A secret government agency recruits some of th...            David Ayer   
..                                                 ...                   ...   
993  While still out to destroy the evil Umbrella C...    Paul W.S. Anderson   
994  3 high school seniors throw a birthday party t...       Nima Nourizadeh   
996  Three American college students studying abroa...              Eli Roth   
997  Romantic sparks occur between two dance studen...            Jon M. Chu   
999  A stuffy businessman finds himself trapped ins...      Barry Sonnenfeld   

                                                Actors  Year  \
0    Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   
1    Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
2    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016   
3    Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016   
4    Will Smith, Jared Leto, Margot Robbie, Viola D...  2016   
..                                                 ...   ...   
993  Milla Jovovich, Ali Larter, Wentworth Miller,K...  2010   
994  Thomas Mann, Oliver Cooper, Jonathan Daniel Br...  2012   
996  Lauren German, Heather Matarazzo, Bijou Philli...  2007   
997  Robert Hoffman, Briana Evigan, Cassie Ventura,...  2008   
999  Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

     Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  
0                  121     8.1  757074              333.13       76.0  
1                  124     7.0  485820              126.46       65.0  
2                  117     7.3  157606              138.12       62.0  
3                  108     7.2   60545              270.32       59.0  
4                  123     6.2  393727              325.02       40.0  
..                 ...     ...     ...                 ...        ...  
993                 97     5.9  140900               60.13       37.0  
994                 88     6.7  164088               54.72       48.0  
996                 94     5.5   73152               17.54       46.0  
997                 98     6.2   70699               58.01       50.0  
999                 87     5.3   12435               19.64       11.0  

[838 rows x 12 columns]

In [191]:
# Drop all columns containing missing data
data.dropna(axis=1)

Rank                    Title                     Genre  \
0       1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1       2               Prometheus  Adventure,Mystery,Sci-Fi   
2       3                    Split           Horror,Thriller   
3       4                     Sing   Animation,Comedy,Family   
4       5            Suicide Squad  Action,Adventure,Fantasy   
..    ...                      ...                       ...   
995   996     Secret in Their Eyes       Crime,Drama,Mystery   
996   997          Hostel: Part II                    Horror   
997   998   Step Up 2: The Streets       Drama,Music,Romance   
998   999             Search Party          Adventure,Comedy   
999  1000               Nine Lives     Comedy,Family,Fantasy   

                                           Description              Director  \
0    A group of intergalactic criminals are forced ...            James Gunn   
1    Following clues to the origin of mankind, a te...          Ridley Scott   
2    Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3    In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4    A secret government agency recruits some of th...            David Ayer   
..                                                 ...                   ...   
995  A tight-knit team of rising investigators, alo...             Billy Ray   
996  Three American college students studying abroa...              Eli Roth   
997  Romantic sparks occur between two dance studen...            Jon M. Chu   
998  A pair of friends embark on a mission to reuni...        Scot Armstrong   
999  A stuffy businessman finds himself trapped ins...      Barry Sonnenfeld   

                                                Actors  Year  \
0    Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   
1    Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
2    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016   
3    Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016   
4    Will Smith, Jared Leto, Margot Robbie, Viola D...  2016   
..                                                 ...   ...   
995  Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...  2015   
996  Lauren German, Heather Matarazzo, Bijou Philli...  2007   
997  Robert Hoffman, Briana Evigan, Cassie Ventura,...  2008   
998  Adam Pally, T.J. Miller, Thomas Middleditch,Sh...  2014   
999  Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

     Runtime (Minutes)  Rating   Votes  
0                  121     8.1  757074  
1                  124     7.0  485820  
2                  117     7.3  157606  
3                  108     7.2   60545  
4                  123     6.2  393727  
..                 ...     ...     ...  
995                111     6.2   27585  
996                 94     5.5   73152  
997                 98     6.2   70699  
998                 93     5.6    4881  
999                 87     5.3   12435  

[1000 rows x 10 columns]

Here we can clearly see that 2 columns with missing data are dropped

we can use <b>thresh</b> parameter to specify the minimum number of non null values for the column/row to be held without dropping

In [205]:
data.dropna(axis=0, thresh = 10)

Rank                    Title                     Genre  \
0       1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1       2               Prometheus  Adventure,Mystery,Sci-Fi   
2       3                    Split           Horror,Thriller   
3       4                     Sing   Animation,Comedy,Family   
4       5            Suicide Squad  Action,Adventure,Fantasy   
..    ...                      ...                       ...   
995   996     Secret in Their Eyes       Crime,Drama,Mystery   
996   997          Hostel: Part II                    Horror   
997   998   Step Up 2: The Streets       Drama,Music,Romance   
998   999             Search Party          Adventure,Comedy   
999  1000               Nine Lives     Comedy,Family,Fantasy   

                                           Description              Director  \
0    A group of intergalactic criminals are forced ...            James Gunn   
1    Following clues to the origin of mankind, a te...          Ridley Scott   
2    Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3    In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4    A secret government agency recruits some of th...            David Ayer   
..                                                 ...                   ...   
995  A tight-knit team of rising investigators, alo...             Billy Ray   
996  Three American college students studying abroa...              Eli Roth   
997  Romantic sparks occur between two dance studen...            Jon M. Chu   
998  A pair of friends embark on a mission to reuni...        Scot Armstrong   
999  A stuffy businessman finds himself trapped ins...      Barry Sonnenfeld   

                                                Actors  Year  \
0    Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   
1    Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
2    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016   
3    Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016   
4    Will Smith, Jared Leto, Margot Robbie, Viola D...  2016   
..                                                 ...   ...   
995  Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...  2015   
996  Lauren German, Heather Matarazzo, Bijou Philli...  2007   
997  Robert Hoffman, Briana Evigan, Cassie Ventura,...  2008   
998  Adam Pally, T.J. Miller, Thomas Middleditch,Sh...  2014   
999  Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

     Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  
0                  121     8.1  757074              333.13       76.0  
1                  124     7.0  485820              126.46       65.0  
2                  117     7.3  157606              138.12       62.0  
3                  108     7.2   60545              270.32       59.0  
4                  123     6.2  393727              325.02       40.0  
..                 ...     ...     ...                 ...        ...  
995                111     6.2   27585                 NaN       45.0  
996                 94     5.5   73152               17.54       46.0  
997                 98     6.2   70699               58.01       50.0  
998                 93     5.6    4881                 NaN       22.0  
999                 87     5.3   12435               19.64       11.0  

[1000 rows x 12 columns]


Here none of the rows are dropped because data is missing only in 2 columns

In [210]:
# Use drop function to drop columns
data.drop('Metascore', axis=1).head()

Rank                    Title                     Genre  \
0     1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1     2               Prometheus  Adventure,Mystery,Sci-Fi   
2     3                    Split           Horror,Thriller   
3     4                     Sing   Animation,Comedy,Family   
4     5            Suicide Squad  Action,Adventure,Fantasy   

                                         Description              Director  \
0  A group of intergalactic criminals are forced ...            James Gunn   
1  Following clues to the origin of mankind, a te...          Ridley Scott   
2  Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3  In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4  A secret government agency recruits some of th...            David Ayer   

                                              Actors  Year  Runtime (Minutes)  \
0  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014                121   
1  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012                124   
2  James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016                117   
3  Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016                108   
4  Will Smith, Jared Leto, Margot Robbie, Viola D...  2016                123   

   Rating   Votes  Revenue (Millions)  
0     8.1  757074              333.13  
1     7.0  485820              126.46  
2     7.3  157606              138.12  
3     7.2   60545              270.32  
4     6.2  393727              325.02

Here we can see that 'Metascore' column is dropped completely.

We know that there are some records where the Revenue is null.

We can impute these null values with mean Revenue,

fillna( ) --> function used to fill null values with specified values

In [213]:
revenue_mean = data['Revenue (Millions)'].mean()
print("The mean revenue is: ", revenue_mean)

The mean revenue is:  82.95637614678898


In [215]:
# We can fill the null values with this mean revenue
data['Revenue (Millions)'].fillna(revenue_mean, inplace=True)


inplace = True signifies that the changes be made permanently in the dataset.

In [217]:
data.isnull().sum()

Rank                   0
Title                  0
Genre                  0
Description            0
Director               0
Actors                 0
Year                   0
Runtime (Minutes)      0
Rating                 0
Votes                  0
Revenue (Millions)     0
Metascore             64
dtype: int64

# Apply( ) Functions

* The apply( ) comes in handy when we want to apply any function to the dataset
* apply( ) function returns a value after passing each row of the dataframe to some function
* The function can be built-in or user defined
For example, if we want to classify the movies based on their ratings, we can define a function to do so and then apply the function to the dataframe as shown below

In [219]:
# Classify movies based on ratings
def rating_group(rating):
    if rating >= 7.5:
        return 'Good'
    elif rating >= 6.0:
        return 'Average'
    else:
        return 'Bad'

In [221]:
# Lets apply this function on our movies data
# creating a new variable in the dataset to hold the rating category
data['Rating_category'] = data['Rating'].apply(rating_group)

In [222]:
data[['Title','Director','Rating','Rating_category']].head(5)

Title              Director  Rating Rating_category
0  Guardians of the Galaxy            James Gunn     8.1            Good
1               Prometheus          Ridley Scott     7.0         Average
2                    Split    M. Night Shyamalan     7.3         Average
3                     Sing  Christophe Lourdelet     7.2         Average
4            Suicide Squad            David Ayer     6.2         Average

# Pivot table

* Pivot table is another useful functionality when it comes to data analysis

* Remeber the pivot tables in excel? Pandas can be used to create such pivot tables as well

* Pivot table takes the column wise data as input and groups the rows to create summary of data

* The pivot table should be created based on a particular column, index and aggregation function to be applied on top of the data

In [229]:
data.pivot_table('Revenue (Millions)', index='Director', columns='Year').head(5)

Year                   2006    2007    2008  2009    2010  2011   2012  2013  \
Director                                                                       
Aamir Khan              NaN    1.20     NaN   NaN     NaN   NaN    NaN   NaN   
Abdellatif Kechiche     NaN     NaN     NaN   NaN     NaN   NaN    NaN   2.2   
Adam Leon               NaN     NaN     NaN   NaN     NaN   NaN    NaN   NaN   
Adam McKay           148.21     NaN  100.47   NaN  119.22   NaN    NaN   NaN   
Adam Shankman           NaN  118.82     NaN   NaN     NaN   NaN  38.51   NaN   

Year                 2014   2015       2016  
Director                                     
Aamir Khan            NaN    NaN        NaN  
Abdellatif Kechiche   NaN    NaN        NaN  
Adam Leon             NaN    NaN  82.956376  
Adam McKay            NaN  70.24        NaN  
Adam Shankman         NaN    NaN        NaN

In [261]:
data.pivot_table('Revenue (Millions)', index='Director', columns='Year')

Year                   2006    2007    2008    2009    2010   2011   2012  \
Director                                                                    
Aamir Khan              NaN    1.20     NaN     NaN     NaN    NaN    NaN   
Abdellatif Kechiche     NaN     NaN     NaN     NaN     NaN    NaN    NaN   
Adam Leon               NaN     NaN     NaN     NaN     NaN    NaN    NaN   
Adam McKay           148.21     NaN  100.47     NaN  119.22    NaN    NaN   
Adam Shankman           NaN  118.82     NaN     NaN     NaN    NaN  38.51   
...                     ...     ...     ...     ...     ...    ...    ...   
Xavier Dolan            NaN     NaN     NaN     NaN     NaN    NaN    NaN   
Yimou Zhang             NaN     NaN     NaN     NaN     NaN    NaN    NaN   
Yorgos Lanthimos        NaN     NaN     NaN    0.11     NaN    NaN    NaN   
Zack Snyder          210.59     NaN     NaN  107.50     NaN  36.38    NaN   
Zackary Adler           NaN     NaN     NaN     NaN     NaN    NaN    NaN   

Year                   2013  2014   2015        2016  
Director                                              
Aamir Khan              NaN   NaN    NaN         NaN  
Abdellatif Kechiche    2.20   NaN    NaN         NaN  
Adam Leon               NaN   NaN    NaN   82.956376  
Adam McKay              NaN   NaN  70.24         NaN  
Adam Shankman           NaN   NaN    NaN         NaN  
...                     ...   ...    ...         ...  
Xavier Dolan            NaN  3.49    NaN   82.956376  
Yimou Zhang             NaN   NaN    NaN   45.130000  
Yorgos Lanthimos        NaN   NaN   8.70         NaN  
Zack Snyder          291.02   NaN    NaN  330.250000  
Zackary Adler           NaN   NaN   6.53         NaN  

[644 rows x 11 columns]

In [262]:
data[(data["Director"] == "Adam McKay") & (data["Year"] == 2006)]

Rank                                        Title                Genre  \
935   936  Talladega Nights: The Ballad of Ricky Bobby  Action,Comedy,Sport   

                                           Description    Director  \
935  #1 NASCAR driver Ricky Bobby stays atop the he...  Adam McKay   

                                                Actors  Year  \
935  Will Ferrell, John C. Reilly, Sacha Baron Cohe...  2006   

     Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  \
935                108     6.6  137502              148.21       66.0   

    Rating_category  
935         Average